Check for ENOENT in the unlink SimProcedure

embg opened the pull request:
Fixes #1633 

* Adds code in `procedures/linux_kernel/unlink.py` to check whether the entity being deleted exists in the filesystem before attempting to delete it. If the entity does not exist, `state.libc.errno` is set to `ENOENT`.

* Adds a test to check that `state.libc.errno` is set correctly by the unlink syscall.

rhelmot commented:
1) sorry for taking this long to get back to you on this
2) You need to _return_ the value of `state.libc.ret_errno()`. It will return -1 if you're emulating a library function and return -errno if you're emulating a syscall that way.
3) What on earth is going on with the testcase? Why do you need all the test code to be inside a simprocedure? Is the global mock patch of libc.errno for the test patch a good idea? We run tests multiprocessed, so this might affect other tests running concurrently...

embg commented:
@rhelmot Thank you for reviewing! I've been busy with classes anyway, so no worries on the delay! 

Re (2): That's an easy fix, I'll change it to return the value of `state.libc.ret_errno()`.

Re (3): Initially, I tried just running the unlink SimProcedure on a blank SimState. That didn't work because `state.libc.errno_location == None` for blank SimStates (since `state.libc` hasn't been initialized yet). The only way I could find to fix the problem was to load in an actual binary, fauxware, and then hook my test to the main of that binary. Since main is called after `state.libc` is initialized, this ensures `state.errno_location` is already set. 

Do you know of a way to initialize `state.libc` without loading in a full binary? If so, that would make the test much simpler. Alternatively, I could make a separate PR to the angr/binaries repo with an appropriate test binary (instead of inline_calling from the fauxware binary).

rhelmot commented:
errno_location is initialized by `__libc_start_main`. You can either run the simprocedure or copy the relevant initialization code:

```python
    def _initialize_errno(self):
        malloc = angr.SIM_PROCEDURES['libc']['malloc']
        errno_loc = self.inline_call(malloc, self.state.arch.bytes).ret_expr

        self.state.libc.errno_location = errno_loc
        self.state.memory.store(errno_loc, self.state.solver.BVV(0, self.state.arch.bits))
```

you can also just set errno_location to some address yourself. it doesn't particularly care where it points and angr treats all memory as mapped by default.

For the mock, you're right, but I live perpetually in fear of nose's multiprocess and desperately want to be as safe as possible with it.

embg commented:
Sorry for the long delay! I simplified the testcase along the lines you proposed and fixed the SimProcedure so that it returns the value of `state.libc.ret_errno()`. Please let me know if there are any other issues.

rhelmot commented:
Looks good. I'm going to fix that lint test for you.

embg commented:
Thanks! I actually tried to do another push to add a newline but git wouldn't count it as a change for some reason :)
